STYLE.md

```diff
@@ -0,0 +1,4068 @@
+    - [Omit Zero Value Fields in Structs](#omit-zero-value-fields-in-structs)
+    - [Use `var` for Zero Value Structs](#use-var-for-zero-value-structs)
+    - [Initializing Struct References](#initializing-struct-references)
+  - [Initializing Maps](#initializing-maps)
+  - [Format Strings outside Printf](#format-strings-outside-printf)
+  - [Naming Printf-style Functions](#naming-printf-style-functions)
+- [Patterns](#patterns)
+  - [Test Tables](#test-tables)
+  - [Functional Options](#functional-options)
+- [Linting](#linting)
+
+## Introduction
+
+Styles are the conventions that govern our code. The term style is a bit of a
+misnomer, since these conventions cover far more than just source file
+formatting—gofmt handles that for us.
+
+The goal of this guide is to manage this complexity by describing in detail the
+Dos and Don'ts of writing Go code at Uber. These rules exist to keep the code
+base manageable while still allowing engineers to use Go language features
+productively.
+
+This guide was originally created by [Prashant Varanasi](https://github.com/prashantv) and [Simon Newton](https://github.com/nomis52) as
+a way to bring some colleagues up to speed with using Go. Over the years it has
+been amended based on feedback from others.
+
+This documents idiomatic conventions in Go code that we follow at Uber. A lot
+of these are general guidelines for Go, while others extend upon external
+resources:
+
+1. [Effective Go](https://go.dev/doc/effective_go)
+2. [Go Common Mistakes](https://go.dev/wiki/CommonMistakes)
+3. [Go Code Review Comments](https://go.dev/wiki/CodeReviewComments)
+
+We aim for the code samples to be accurate for the two most recent minor versions
+of Go [releases](https://go.dev/doc/devel/release).
+
+All code should be error-free when run through `golint` and `go vet`. We
+recommend setting up your editor to:
+
+- Run `goimports` on save
+- Run `golint` and `go vet` to check for errors
+
+You can find information in editor support for Go tools here:
+https://go.dev/wiki/IDEsAndTextEditorPlugins
+
+## Guidelines
+
+### Pointers to Interfaces
+
+You almost never need a pointer to an interface. You should be passing
+interfaces as values—the underlying data can still be a pointer.
+
+An interface is two fields:
+
+1. A pointer to some type-specific information. You can think of this as
+   "type."
+2. Data pointer. If the data stored is a pointer, it’s stored directly. If
+   the data stored is a value, then a pointer to the value is stored.
+
+If you want interface methods to modify the underlying data, you must use a
+pointer.
+
+### Verify Interface Compliance
+
+Verify interface compliance at compile time where appropriate. This includes:
+
+- Exported types that are required to implement specific interfaces as part of
+  their API contract
+- Exported or unexported types that are part of a collection of types
+  implementing the same interface
+- Other cases where violating an interface would break users
```

Comment:
I feel that having the unexported interfaces is a bit skewed in the first place. 

Ideally those interfaces should be exposed already publicly on the dependency package.

For example, if the p2p server exposes some pubsub methods through an interface, it should also generate the p2p mocks and we should consume those everywhere in the codebase where this dependency is used (instead of duplicating those interfaces endlessly, and generating redundant mocks). It generates less points of friction for testing and for handling the dependency injection.

In some cases indeed subsets of interfaces are needed but I'm not sure why not to just compose the interfaces together at the dependency level, this is not unheard of.

Another point for public interfaces is allowing us to properly do black-box testing which I feel is way superior to white-box. Ideally, white-box should be resorted to only in exceptional cases, but otherwise tests shouldn't leak package internals. Currently most of the tests don't have namespace separation and I feel it causes a lot of mess when asserting state. A lot of the necessary helpers can be put in an `export_test.go` [file](https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd) (which is a bit hacky in the first place, but at least puts a lid on hackiness).

Reply:
And back to the conversation about interface compliance - it would be nice to have, but we could also change it to compliance with external interfaces only for now, it is fine by me.

Reply:
The implementation check `var _ MyInterface = MyType{}` doesn't add anything to our codebase, also without it we quickly notice that a type that should implement an interface doesn't any more because at the point of its use it will cause a compilation error.

We are not providing a library to be used by others the interfaces we use exist for decoupling and mocking. The reason we define interfaces at the user side is because the user decides which subset of the service they want to use and/or mock.

If the implementation changes its signature we have to update all uses anyway, doesn't matter if the interface is on the user or the provider side. I don't mind trying more aggressively to go to black box tests, but even then I see no need to define interfaces at the implementation side - we want to avoid dependencies between packages. Ideally only `node` has a dependency on (almost) everything while all other packages have dependencies on (almost) nothing.

We are right now struggling with refactorings because interfaces were defined on the implemenation side: In the `malfeasance` package used to be the interface and types for all packages creating malfeasance proofs, but those proofs contain types of those other packages:

- `activation` creates a `malfeasance` proof, inside there is a `activation.ActivationTx`
- we now have a cyclic dependency between `activation` and `malfeasance`
- the "solution" was to create more packages: `malfeasance` depends on `malfeasance/wire`, `activation/wire` and `activation` depends on `malfeasance`, `malfeasance/wire` and `activation/wire`
- its now 4 packages instead of 2 just to get around the issue of that activation depends on malfeasance for publishing proofs and malfeasance depends on activation to validate them

Reply:
Ok point made :) I will amend